OOD Attempt 1

plopeziv opened the pull request.

plopeziv committed:
OOD Attempt 1

digitopolis commented on js-jest/src/gilded_rose.js:
```diff
@@ -9,201 +9,141 @@ class Item {
+  }
+
+  updateQuality() {
+    for (let i = 0; i < this.items.length; i++) {
+      this.items[i].updateItem()
+    }
```
This is 100% valid - but now might be a good opportunity to check out some other modern JS iterating methods:
[for...of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of)
[forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)
[good ol' map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)

One nice thing about them is that they can sometimes help make things a little more human-readable. For example:
```suggestion
    for (const item of this.items) {
      item.updateItem()
    }
```

or

```suggestion
    this.items.forEach(item => item.updateItem())
```

Again – absolutely nothing wrong with the way you did it though!

plopeziv replied:
Features implemented in newest update

digitopolis reviewed (changes requested):
This looks good - much more open-closed. The only thing I noticed is that we're not decrementing the `sellIn` value, as is expected from the original specs/implementation. Right now, items will increase/decrease in quality (where appropriate) but will never approach their sell by date.

plopeziv committed:
Update 2

Number of tests reduced as they were found redundant after OOD operation

Tests refactored to support OOD operations

Use of iterator for cleaner code

SellIn degredation bug fixed

davetorre commented on js-jest/src/gilded_rose.js:
```diff
@@ -9,201 +9,142 @@ class Item {
+  updateQuality() {
+    this.items.forEach(item => {
+      item.updateItem()
+      item.sellIn -= 1})
```
The LegendaryItem not conforming to this abstraction is kind of annoying right?
To me, it feels weird to have to set its quality to 0 each time `updateItem` is called. So I'd probably move this logic that decrements `sellIn` out of here and into each item type's class.
Also, I did a quick look at the requirements and I think Sulfuras is supposed to have a quality of 80 always.

davetorre replied:
Talked IRL and agreed that `sellIn` is ok to leave as-is at this point. Still want to fix the 80 though.

davetorre commented on js-jest/test/gilded_rose.test.js:
```diff
@@ -1,20 +1,39 @@
+const {Shop, Item, NormalItem, ConjuredItem, AgedBrie,
+  LegendaryItem, ConcertTickets} = require("../src/gilded_rose");
 
 test("First Normal Item Test ", function() {
```
Now that you've done the refactor to make this Open-Closed, these tests seem more like integration tests to me and less like unit tests. I say this because they test not only the Shop class, but also the functionality you've encapsulated inside each of the item type classes.

davetorre reviewed (changes requested):
I think the tests are actually fine to merge in as they are. I'm requesting changes because of my other comment about LegendaryItem.
Would be cool to hear what you think about the testing comment though, since it goes along with the testing pyramid idea.

plopeziv committed:
Separate items into classes and create item tests
